README.md
# lifecycle_state

A flutter package similar to Android&#x27;s lifecycle.  
给flutter添加了类似于Android生命周期的方法，其中`LifecycleState`适用于普通页面，只要把我们用的State替换成LifecycleState即可,`LifecycleInnerState`适用于PageView中的页面

# Getting Started
### 导入`lifecycle_state`
```
lifecycle_state:
    git:
      url: "https://github.com/tianyu704/lifecycle_state.git"
```
or
```
lifecycle_state: ^0.0.1
```
### MaterialApp中添加`navigatorObservers: [routeObserver]`
```
import 'package:lifecycle_state/lifecycle_state.dart';

@override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Test',
      navigatorObservers: [routeObserver],
      ...
    );
  }
```  
routeObserver 是在lifecycle_state中定义的

### 页面中的使用
1. 直接继承`LifecycleState`

```
class TestPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    // TODO: implement createState
    return _TestPageState();
  }
}

class _TestPageState extends LifecycleState<TestPage> {
  PageController _controller;

  @override
  void onCreate() {
    // TODO: implement onCreate
    super.onCreate();
    _controller = PageController();
//    log("onCreate");
  }

  @override
  Widget build(BuildContext context) {
    // TODO: implement buildWidget
    return Scaffold(
      body: PageView(
        controller: _controller,
        onPageChanged: (index) {
          eventBus.fire(LifecycleInnerEvent(index, "tag"));
        },
        children: <Widget>[
          ItemPage(0, "tag"),
          ItemPage(1, "tag"),
          ItemPage(2, "tag"),
        ],
      ),
      floatingActionButton: FloatingActionButton(onPressed: () {
        Navigator.of(context).push(MaterialPageRoute(builder: (context) {
          return SecondPage();
        }));
      }),
    );
  }

  @override
  void onResume() {
    // TODO: implement onResume
    super.onResume();
    log("onResume");
  }

  @override
  void onPause() {
    // TODO: implement onPause
    super.onPause();
    log("onPause");
  }

  @override
  void onDestroy() {
    // TODO: implement onDestroy
    super.onDestroy();
    log("onDestroy");
  }

  @override
  void onLoadData() {
    // TODO: implement onLoadData
    super.onLoadData();
    log("onLoadData");
  }

  @override
  void onBackground() {
    // TODO: implement onBackground
    super.onBackground();
    log("onBackground");
  }

  @override
  void onForeground() {
    // TODO: implement onForeground
    super.onForeground();
    log("onForeground");
  }
}
```
2. 如果要在PageView中使用，flutter默认每次切换page，旧的page是销毁的，如果你的页面可以这样就可以直接用LifecycleState，但如果你不想每次都销毁重建，就用`LifecycleInnerState`，但这个稍微麻烦的是得手动通知页面状态变化，这里暂时使用`event_bus`通信，就是在PageView的onPageChanged中导包并调用`eventBus.fire(LifecycleInnerEvent(index, "tag"));`,"tag"是用来标记每个pageview的，避免一个pageview改变，通知到别的pageview。`LifecycleInnerState`中加入了`AutomaticKeepAliveClientMixin`防止页面销毁,你自己的类中就不要加了
```
import 'package:schulte_grid_flutter/src/lifecycle_inner_state.dart';
...
PageView(
        controller: _controller,
        onPageChanged: (index) {
          eventBus.fire(LifecycleInnerEvent(index, "tag"));
        },
        children: <Widget>[
          ItemPage(0),
          ItemPage(1),
          ItemPage(2),
        ],
      ),
...
```
```
class ItemPage extends StatefulWidget {
  final int index;
  ItemPage(this.index);

  @override
  State<StatefulWidget> createState() {
    // TODO: implement createState
    return _ItemPageState();
  }
}

class _ItemPageState extends LifecycleInnerState<ItemPage> {
  @override
  Widget build(BuildContext context) {
    // TODO: implement build
    return Scaffold(
      body: Center(
        child: Text("${widget.index}"),
      ),
    );
  }

  @override
  int setPosition() {
    // TODO: implement setPosition
    return widget.index;
  }

  @override
  void onLoadData() {
    // TODO: implement onLoadData
    super.onLoadData();
    log("onLoadData");
  }

  @override
  void onResume() {
    // TODO: implement onResume
    super.onResume();
    log("onResume");
  }

  @override
  void onPause() {
    // TODO: implement onPause
    super.onPause();
    log("onPause");
  }

  @override
  void onDestroy() {
    // TODO: implement onDestroy
    super.onDestroy();
    log("onDestroy");
  }

  @override
  void onCreate() {
    // TODO: implement onCreate
    super.onCreate();
    log("onCreate");
  }

  @override
  // TODO: implement wa
  //  ntKeepAlive
  bool get wantKeepAlive => true;

  @override
  String setTag() {
    // TODO: implement setTag
    return "tag";
  }
}
```
参考：  
1. LifecycleState参考于 https://github.com/tinyvampirepudge/flutter_lifecycle_state ，但是他的这个我用着是有bug的，当跳转到继承原生State的页面中再回来时，方法调用会出错，有的生命周期方法会不调用了  
2. LifecycleInnerState参考于 https://github.com/385841539/flutter_BaseWidget 中的一部分，他这个有一个通用的页面生命周期的基类，采用一个map记录了所有页面，然后通过判断当前页面是栈顶或第二个来调用相应的生命周期方法，因此你必须所有页面都继承他的基类，如果跳转到不继承基类的页面中那么那些方法调用就是错误的

综合这两个的优缺点自己改了一个版本，在想用的时候就用`LifecycleState`或`fecycleInnerState`，不想用也没关系,不影响用的页面




CHANGELOG.md
## [0.0.1] - First

* publish

## [0.0.2] - Update
* update event_bus version

test/lifecycle_state_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:lifecycle_state/lifecycle_state.dart';

void main() {
//  test('adds one to input values', () {
//    final calculator = Calculator();
//    expect(calculator.addOne(2), 3);
//    expect(calculator.addOne(-7), -6);
//    expect(calculator.addOne(0), 1);
//    expect(() => calculator.addOne(null), throwsNoSuchMethodError);
//  });
}


lib/src/route_name_util.dart
import 'package:flutter/material.dart';

///
/// Create by Hugo.Guo
/// Date: 2019-08-09
///
class RouteNameUtil {
  /// name+id
  static String getRouteNameWithId(BuildContext context) {
    if (context == null) {
      return null;
    }
    String name = context.widget?.toString();
    if (name == null) {
      return null;
    }
    String contextString = context.toString();
    if (contextString == null) {
      return name;
    }
    String id;
    try {
      id = contextString.substring(
          contextString.indexOf("#"), contextString.indexOf("(lifecycle"));
    } catch (e) {
      id = null;
    }
    if (id == null) {
      return name;
    }
    return name + id;
  }

  /// name
  static String getRouteName(BuildContext context) {
    if (context == null) {
      return null;
    }
    String name = context.widget?.toString();
    if (name == null) {
      return null;
    }
    return name;
  }
}


lib/src/navigator_manger.dart
import 'package:flutter/material.dart';

/// 管理生命周期
class NavigatorManger {
  List<String> _activityStack = new List<String>();

  NavigatorManger._internal();

  static NavigatorManger _singleton = new NavigatorManger._internal();

  //工厂模式
  factory NavigatorManger() => _singleton;

  void addWidget(String className) {
    _activityStack.add(className);
  }

  void removeWidget(String className) {
    _activityStack.remove(className);
  }

  bool isTopPage(String className) {
    if (_activityStack == null) {
      return false;
    }
    try {
      return className == _activityStack[_activityStack.length - 1];
    } catch (exception) {
      return false;
    }
  }

  bool isSecondTop(String className) {
    if (_activityStack == null) {
      return false;
    }
    try {
      return className == _activityStack[_activityStack.length - 2];
    } catch (exception) {
      return false;
    }
  }
}


lib/src/lifecycle_inner_state.dart
import 'package:flutter/material.dart';
import 'route_name_util.dart';
import 'package:event_bus/event_bus.dart';

///
/// Create by Hugo.Guo
/// Date: 2019-08-08
///
abstract class LifecycleInnerState<T extends StatefulWidget> extends State<T>
    with AutomaticKeepAliveClientMixin {
  bool _isResume = false;
  bool _isPause = false;
  int _position;
  String _tag;
  String _className;

  get className => _className;

  int setPosition();

  String setTag();

  @override
  void initState() {
    // TODO: implement initState
    _className = RouteNameUtil.getRouteNameWithId(context);
    _position = setPosition();
    _tag = setTag();
    onCreate();
    onResume();
    onLoadData();
    eventBus.on<LifecycleInnerEvent>().listen((event) {
      if (event.tag == _tag) {
        if (_position == event.currentPage) {
          if (!_isResume) {
            onResume();
          }
        } else {
          if (!_isPause) {
            onPause();
          }
        }
      }
    });
    super.initState();
  }

  @override
  void dispose() {
    // TODO: implement dispose
    onDestroy();
    super.dispose();
  }

  ///初始化一些变量 相当于 onCreate ， 放一下 初始化数据操作
  void onCreate() {}

  ///相当于onResume, 只要页面来到栈顶， 都会调用此方法
  void onResume() {
    _isResume = true;
    _isPause = false;
  }

  /// 加载数据,网络请求等
  void onLoadData() {}

  ///页面被覆盖,暂停
  void onPause() {
    _isResume = false;
    _isPause = true;
  }

  ///页面销毁的方法
  void onDestroy() {}

  /// log输出
  void log(String log) {
    debugPrint(_className + "---->" + log);
  }

  @override
  // TODO: implement wantKeepAlive
  bool get wantKeepAlive => keepAlive();

  /// 子页面可以决定是否销毁,默认不销毁
  bool keepAlive() {
    return true;
  }
}

class LifecycleInnerEvent extends Notification {
  final int currentPage;
  final String tag;

  LifecycleInnerEvent(this.currentPage, this.tag);
}

EventBus eventBus = EventBus();

onPageChanged(index, tag) {
  eventBus.fire(LifecycleInnerEvent(index, tag));
}


lib/src/lifecycle_state.dart
import 'package:flutter/material.dart';

///
/// Create by Hugo.Guo
/// Date: 2019-08-09
///
/// Register the RouteObserver as a navigation observer.
final RouteObserver<PageRoute> routeObserver = RouteObserver<PageRoute>();

abstract class LifecycleState<T extends StatefulWidget> extends State<T>
    with WidgetsBindingObserver, RouteAware {
  String tag = T.toString();

  // 参照State中写法，防止子类获取不到正确的widget。
  T get widget => super.widget;

  //是否在栈顶
  bool _isTop = false;

  @override
  void initState() {
    // TODO: implement initState
    super.initState();
    onCreate();
    onResume();
    onLoadData();
    WidgetsBinding.instance.addObserver(this);
  }

  @override
  void deactivate() {
    // TODO: implement deactivate
    super.deactivate();
//    log("deactivate");
  }

  @override
  void didChangeAppLifecycleState(AppLifecycleState state) {
//    log("AppLifecycleState:$state");
    // TODO: implement didChangeAppLifecycleState
    super.didChangeAppLifecycleState(state);
    switch (state) {
      case AppLifecycleState.resumed:
        if (_isTop) {
          onForeground();
          onResume();
        }
        break;
      case AppLifecycleState.inactive:
        break;
      case AppLifecycleState.paused:
        if (_isTop) {
          onBackground();
          onPause();
        }
        break;
      case AppLifecycleState.detached:
        break;
      default:
        break;
    }
  }

  @override
  void didChangeDependencies() {
    super.didChangeDependencies();
//    log("didChangeDependencies");
    routeObserver.subscribe(this, ModalRoute.of(context));
  }

  @override
  void dispose() {
    // TODO: implement dispose
    onDestroy();
    WidgetsBinding.instance.removeObserver(this);
    routeObserver.unsubscribe(this);
    super.dispose();
  }

  @override
  void didPush() {
    // TODO: implement didPush
    super.didPush();
//    log("didPush");
    _isTop = true;
  }

  @override
  void didPushNext() {
    // TODO: implement didPushNext
    super.didPushNext();
//    log("didPushNext");
    onPause();
    _isTop = false;
  }

  @override
  void didPop() {
    // TODO: implement didPop
    super.didPop();
//    log("didPop");
    onPause();
  }

  @override
  void didPopNext() {
    // TODO: implement didPopNext
    super.didPopNext();
//    log("didPopNext");
    onResume();
    _isTop = true;
  }

  /// 用于让子类去实现的初始化方法
  void onCreate() {
//    log("onCreate");
  }

  /// 用于让子类去实现的不可见变为可见时的方法
  void onResume() {
//    log("onResume");
  }

  ///加载数据
  void onLoadData() {
//    log("onLoadData");
  }

  /// 用于让子类去实现的可见变为不可见时调用的方法。
  void onPause() {
//    log("onPause");
  }

  /// 用于让子类去实现的销毁方法。
  void onDestroy() {
//    log("onDestroy");
  }

  /// app切回到后台
  void onBackground() {
//    log("onBackground");
  }

  /// app切回到前台
  void onForeground() {
//    log("onForeground");
  }

  log(String log) {
    debugPrint('$tag --> $log');
  }
}
